// @workadventure/messages/dist/libs/room-api-clients/google/protobuf/field_mask.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.10.1
//   protoc               v4.25.1
// source: google/protobuf/field_mask.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.FieldMask = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "google.protobuf";
function createBaseFieldMask() {
    return { paths: [] };
}
exports.FieldMask = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        for (const v of message.paths) {
            writer.uint32(10).string(v);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFieldMask();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.paths.push(reader.string());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            paths: typeof object === "string"
                ? object.split(",").filter(globalThis.Boolean)
                : globalThis.Array.isArray(object === null || object === void 0 ? void 0 : object.paths)
                    ? object.paths.map(globalThis.String)
                    : [],
        };
    },
    toJSON(message) {
        return message.paths.join(",");
    },
    create(base) {
        return exports.FieldMask.fromPartial(base !== null && base !== void 0 ? base : {});
    },
    fromPartial(object) {
        var _a;
        const message = createBaseFieldMask();
        message.paths = ((_a = object.paths) === null || _a === void 0 ? void 0 : _a.map((e) => e)) || [];
        return message;
    },
    wrap(paths) {
        const result = createBaseFieldMask();
        result.paths = paths;
        return result;
    },
    unwrap(message) {
        return message.paths;
    },
};
